### PR TITLE
Refactor `put`

### PR DESCRIPTION
**Pull request by vinamrabenara**

### CODE REVIEW

File: nums/core/array/blockarray.py [hunk 1]
```diff
@@ -363,7 +366,13 @@ def _advanced_single_array_select(self, ss: tuple, axis: int = 0):
 
         src_arr = self
         np_ss = ss
-        ss = self.cm.put(ss)
+        ss = self.cm.put(
+            ss,
+            syskwargs={
+                "grid_entry": 0,
```
Comment (vinamrabenara) [inline nums/core/array/blockarray.py:372]: @elibol Can you check this. I am not sure if the syskwargs are correct here

File: nums/core/array/base.py [hunk 2]
```diff
@@ -172,7 +172,13 @@ def _block_from_other(self, other):
             False,
             self._cm,
         )
-        block.oid = self._cm.put(np.array(other, dtype=self.dtype))
+        block.oid = self._cm.put(
+            np.array(other, dtype=self.dtype),
+            syskwargs={
```
Comment (elibol) [inline nums/core/array/base.py:177]: Let's add a comment here justifying the `syskwargs` of `other` here.

File: nums/core/array/blockarray.py (hunk 1 above)
Comment (elibol) [reply nums/core/array/blockarray.py:372]: Looks good, just set `"grid_shape": 1` so there's no confusion about what's going on. `ss` will be pulled to other nodes as needed.

File: nums/core/array/blockarray.py [hunk 3]
```diff
@@ -509,7 +518,13 @@ def _advanced_single_array_assign(
         src_arr = value
         src_grid_shape = src_arr.grid.grid_shape
         np_ss = ss
-        ss = self.cm.put(ss)
+        ss = self.cm.put(
+            ss,
+            syskwargs={
+                "grid_entry": 0,
+                "grid_shape": src_grid_shape,
```
Comment (elibol) [inline nums/core/array/blockarray.py:525]: Same here, set `"grid_shape": 1`.

File: nums/core/compute/compute_interface.py [hunk 4]
```diff
@@ -147,6 +147,9 @@ def arg_op(
     def reshape(self, arr, shape, syskwargs: Dict):
         raise NotImplementedError()
 
+    def put(self, value):
```
Comment (elibol) [inline nums/core/compute/compute_interface.py:150]: Can you rename this to `identity`?

File: nums/core/compute/compute_manager.py [hunk 5]
```diff
@@ -113,8 +113,13 @@ def __getattribute__(self, name: str):
     # System Interface
     ####################
 
-    def put(self, value: Any):
-        return self.system.put(value)
+    def put(self, value: Any, **kwargs):
+        assert "syskwargs" in kwargs
+        kwargs = kwargs.copy()
+        syskwargs = kwargs["syskwargs"]
+        del kwargs["syskwargs"]
```
Comment (elibol) [inline nums/core/compute/compute_manager.py:120]: Can you also add `assert "options" not in syskwargs`?

File: nums/core/compute/compute_manager.py [hunk 6]
```diff
@@ -113,8 +113,13 @@ def __getattribute__(self, name: str):
...
+        device_id, options = self._process_syskwargs(syskwargs)
```
Comment (elibol) [inline nums/core/compute/compute_manager.py:121]: To be clear, let's add `assert len(options) == 0`.

File: nums/core/systems/system_interface.py [hunk 7]
```diff
@@ -27,9 +27,9 @@ def init(self):
     def shutdown(self):
         raise NotImplementedError()
 
-    def put(self, value: Any):
+    def put(self, args, kwargs, device_id: DeviceID, options: Dict):
```
Comment (elibol) [inline nums/core/systems/system_interface.py:30]: This signature should just be `put(self, value: Any, device_id: DeviceID)`.

**Review by elibol — COMMENTED** [review]
Looks good so far! Left some comments.

**Comment by vinamrabenara** [comment]
Addressed all comments

File: nums/core/array/base.py (hunk 2 above)
Comment (elibol) [reply nums/core/array/base.py:178]: We want to justify the choice of `syskwargs`, namely, `other` will be a literal that we want local to self. Can you update accordingly? Please use punctuation with all comments.

File: nums/core/systems/system_interface.py [hunk 8]
```diff
@@ -27,9 +27,9 @@ def init(self):
     def shutdown(self):
         raise NotImplementedError()
 
-    def put(self, value: Any):
+    def put(self, value: Any, device_id: DeviceID):
         """
-        Put object into system storage.
+        Put object into system storage and force placement on the relevant node
```
Comment (elibol) [inline nums/core/systems/system_interface.py:32]: Please include punctuation.

File: nums/core/systems/systems.py [hunk 9]
```diff
@@ -184,12 +184,17 @@ def warmup_func(n):
                 for _ in range(n):
                     _a = random.randint(0, 1000)
                     _b = random.randint(0, 1000)
-                    _v = self.get(r(self.put(_a), self.put(_b)))
+                    _v = self.get(
```
Comment (elibol) [inline nums/core/systems/systems.py:187]: Can we have this be round-robin over the devices? Including placing `_a`, `_b` on different nodes?

**Review by elibol — COMMENTED** [review]
Looks good! Left very minor comments.

**Review by elibol — APPROVED** [review]
LGTM.

**Pull request merged**